domains/pallets/domain-id/src/lib.rs

```diff
@@ -44,16 +47,17 @@ mod pallet {
     #[pallet::genesis_build]
     impl<T: Config> GenesisBuild<T> for GenesisConfig {
         fn build(&self) {
-            SelfDomainId::<T>::set(
-                self.domain_id
-                    .expect("Genesis config of pallet-domain-id must be set"),
-            );
+            // NOTE: `domain_id` can set to `None` in the `RuntimeGenesisConfig` of chain spec since
+            // the id of a domain instance is allocated during instantiation and can not be known ahead
+            // of time, but it must be reset to the correct value in the genesis storage.
+            SelfDomainId::<T>::set(self.domain_id);
         }
     }
 
     impl<T: Config> Pallet<T> {
         pub fn self_domain_id() -> DomainId {
             SelfDomainId::<T>::get()
+                .expect("self domain-id must set to Some when building the genesis storage")
```

Comment:
Because the `SelfDomainId` must be reset in the genesis storage after the original genesis storage is built, the workflow:
1. The runtime is registered with the genesis storage, where the `SelfDomainId` is none
2. The domain is instantiated with a new domain id, the `SelfDomainId` of the original genesis storage is reset with the new domain id after that the genesis storage is used in the domain instance chain.

Reply:
Hm, then the comment should be something like this:
```
Domain ID must be set during domain instantiation
```

It is neither set to `Some` (it is the key that is missing, not that we store `None`) nor we do this during building of genesis storage, we do this during domain instantiation instead. I think it is helpful to be accurate here to prevent confusion.